_posts/2015-02-25-elasticsearch-and-denormalization.md

```diff
@@ -0,0 +1,273 @@
+ElasticSearch. The gem expects a method called `to_hash`. The hash is converted
+to JSON for each record and used to populate the ElasticSearch index:
+
+[elasticsearch-model]: https://github.com/elasticsearch/elasticsearch-rails/tree/master/elasticsearch-model
+
+```ruby
+class Sample < ActiveRecord::Base
+  include Elasticsearch::Model
+
+  # Building as small hash here is a little smelly but probably OK.
+  # As the hash grows you should start treating it as a separate concern. This
+  # is what we will be doing with our denormalizer (see below).
+  def to_hash
+    {
+                  id: self.id,
+        display_name: display_name,
+          style_name: style.name,
+            style_id: style.id,
+      original_image: original_image,
+          color_name: color.name
+    }
+  end
+  
+  settings(elasticsearch_settings) do
+    mapping do
+      indexes :display_name, type: 'multi_field', fields: {
+        display_name:       { type: "string", analyzer: "snowball" },
+        display_name_exact: { type: "string", index: :not_analyzed }
+      }
+      # etc ...
+    end
+  end
+end
+```
+The [Presenter pattern][presenter-pattern] is a familiar one in Rails: wrap
+objects in a presenter class to keep business logic away from your HAML or ERB.
+Since the process of building up a hash to push into an ElasticSearch index
+is basically the same we decided to use one class to do both jobs. We start with
+a 'denormalizer' class that flattens out the many objects that make up the
+sample view. The `SampleDenormalizer` looks like this:
+
+[presenter-pattern]: http://nithinbekal.com/posts/rails-presenters/ "Nithin Bekal on Presenters in Rails"
+
+```ruby
+class SampleDenormalizer
+
+  def initializer(sample)
+    @sample = sample
+  end
+
+  # This method should be treated like a view - no logic, just attributes
+  def to_hash
+    # this list of methods can get very long but it encourages you to
+    # keep the hash flat and to be deliberate about what you are adding to it
+    %w(id
+       display_name
+       original_image
+       style_id
+       style_name
+       color_name).map{ |method_name|
+        [ method_name, self.send(method_name) ]
+    }.to_h
+  end
+
+  def display_name
+    @sample.display_name
+  end
+
+  def original_image
+    @sample.original_image
+  end
+
+  def color_name
+    @sample.color.name
+  end
+
+  def style_name
+    style.name
+  end
+
+  def style_id
+    style.id
+  end
+
+  private
+  # any methods that aren't included in the hash can be private and should be
+  # memoized for better performance
+  def style
+    @style  ||= @sample.style
+  end
+
+end
+```
+
+You can now use the denormalizer in the model to build the hash for
+ElasticSearch, considerably reducing the complexity of this particular class.
+
+```ruby
+class Sample < ActiveRecord::Base
+  include Elasticsearch::Model
+
+  def to_hash
+    SampleDenormalizer.new(self).to_hash
+  end
+  
+  # mapping here as before ...
+
+end
+```
+
```

Comment:
Maybe this would be a good spot for another sub-title to break up the post a little more. Something to do with how it can also be used for more complex calculated attributes.